services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/events.py

```diff
@@ -248,19 +248,27 @@ async def will_trigger(cls, app: FastAPI, scheduler_data: SchedulerData) -> bool
 
     @classmethod
     async def action(cls, app: FastAPI, scheduler_data: SchedulerData) -> None:
+        app_settings: AppSettings = app.state.settings
         dynamic_sidecar_client = get_dynamic_sidecar_client(app)
         dynamic_sidecar_endpoint = scheduler_data.dynamic_sidecar.endpoint
 
         async with disabled_directory_watcher(
             dynamic_sidecar_client, dynamic_sidecar_endpoint
         ):
-            # below tasks can take a while, running them in parallel
-            await logged_gather(
-                dynamic_sidecar_client.service_restore_state(dynamic_sidecar_endpoint),
+            tasks = [
                 dynamic_sidecar_client.service_pull_output_ports(
                     dynamic_sidecar_endpoint
-                ),
-            )
+                )
+            ]
+            # When enabled no longer downloads state via nodeports
+            # S3 is used to store state paths
+            if not app_settings.DIRECTOR_V2_DEV_FEATURES_ENABLED:
+                tasks.append(
+                    dynamic_sidecar_client.service_restore_state(
+                        dynamic_sidecar_endpoint
+                    )
+                )
+            await logged_gather(*tasks)
```

Comment:
Can you explain where the state is now saved?
Until now, the state was a zip file which had an entry in the file_meta_data table.
Now I guess that since it is going through RClone, we do not have any entry in the database anymore right?
So in order to be able to clean the storage, we need to have a VERY WELL defined definition of where the node states are located.

Reply:
Say for example a service saves it's state in the following folders:
- `/etc/conf/app`
- `/home/user/work`

The state will be saved at the following paths inside S3:

- `BUCKET_NAME/PROJEC_UUID/NODE_UUID/etc_conf_app`
- `BUCKET_NAME/PROJEC_UUID/NODE_UUID/home_user_work`

You end up with 1 directory per state path  in the `BUCKET_NAME/PROJEC_UUID/NODE_UUID` namespace.

Finally you are correct, there are no further entries in the database.

I think the above way of storing the date allows for easy cleanup and data restoring to services, since it is possible to compute the path from the name.

Reply:
just to be on the safe side, what about having a subfolder for these states?
If later we get inputs/outputs the same. is there not a chance of name collision?
Something like ```BUCKET_NAME/PROJECT_UUID/NODE_UUID/states/...```
Just wondering, I guess you already have an idea about it?

Reply:
If you look closely at the directory name, it is derived from the absolute path on the disk by replacing `/` with ´_´ and stripping the leading `_`.
I do not see risks of collisions here.